### PR TITLE
Introduce a GN rule that (explicitly) generates a `dart test` wrapper

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/flutter/flutter/issues/155769.

This is a variant of the approach in https://github.com/flutter/engine/pull/52241, based on feedback from @jakemac53 and @jonahwilliams (who originally sped up `dart test` significantly by using `frontend_server` under the scenes: https://github.com/dart-lang/test/pull/1399), in short:

```gn
# tools/engine_tool/BUILD.gn

import("//flutter/build/dart/internal/gen_dartcli_call.gni")

gen_dartcli_call("tests") {
  args = [ "test" ]
  cwd = "//flutter/tools/engine_tool"
}
```

This stanza, when built (`ninja -C ../out/host_debug flutter/tools/engine_tool:tests`) generates the following file:

```sh
# ../out/host_debug/gen/flutter/tools/engine_tool/tests

set -e

# Needed because if it is set, cd may print the path it changed to.
unset CDPATH

# Store the current working directory.
prev_cwd=$(pwd)

# Set a trap to restore the working directory.
trap 'cd "$prev_cwd"' EXIT

CD_PATH="/Users/matanl/Developer/engine/src/flutter/tools/engine_tool"
if [ -n "$CD_PATH" ]; then
  cd "$CD_PATH"
fi

/Users/matanl/Developer/engine/src/flutter/prebuilts/macos-arm64/dart-sdk/bin/dart "test"
```

In turn, when executed (`../out/host_debug/gen/flutter/tools/engine_tool/tests`) it just runs `dart test`:

```sh
flutter % ../out/host_debug/gen/flutter/tools/engine_tool/tests
Building package executable... 
Built test:test.
00:00 +0: test/test_command_test.dart: test command executes test                                                                                                                                                                                                          
00:00 +3: test/run_command_test.dart: run command invokes flutter run
...
00:00 +117: All tests passed!
```

There is no actual compilation performed by the tool (that is handled implicitly by `dart test`), and as a result neither a `depfile` is needed, nor generating a pre-compiled artifact like a snapshot or AOT elf/assembly. 

---

This work is incomplete, that is, we'd want to properly tag the executable so `et` can find it, and create a wrapper template (i.e. `dart_test`) that tightens things up a bit, but I wanted to show the work at this intermediate step to get feedback before moving forward.

/cc @jonahwilliams, @jtmcdole as well.

### CODE REVIEW

File: build/dart/internal/gen_executable_call.py
Comment (jonahwilliams) [inline build/dart/internal/gen_executable_call.py:31]: dumb question because I get tripped up on this constantly. set -e makes sure that the exit code of this script is non zero if the dart test exit code is non zero right? Can you check that failed tests report non-zero exit codes through the shell script too?

Comment (jtmcdole) [reply build/dart/internal/gen_executable_call.py:31]: And command or pipeline that has an error will exit the script automatically.  Dart tests that fail set errno 1.

File: build/dart/internal/gen_executable_call.py
Comment (jtmcdole) [inline build/dart/internal/gen_executable_call.py:19]: I <3 `rest` vs `args`  +1

**Comment by zanderso** [comment]
Does `dart test` do an implicit `pub get`?

**Comment by matanlurey** [comment]
> Does `dart test` do an implicit `pub get`?

Yes, I believe it does. My thought that given how workspaces work, it shouldn't matter though?

(Side-note, if it _does_ matter, we should make sure to change [`run_tests.py` in the meantime](https://github.com/flutter/engine/blob/feae96859bd733714cd0e51228495c2523fe4fc1/testing/run_tests.py#L918-L925)).

**Comment by jakemac53** [comment]
> Yes, I believe it does. My thought that given how workspaces work, it shouldn't matter though?

Right, it shouldn't matter, but if we really wanted to skip it there might be a flag for that... cc @bkonyi .

**Comment by bkonyi** [comment]
> > Yes, I believe it does. My thought that given how workspaces work, it shouldn't matter though?
> 
> Right, it shouldn't matter, but if we really wanted to skip it there might be a flag for that... cc @bkonyi .

I don't think there's a flag to disable the automatic `pub get` since `dart test` triggers some precompilation steps before running tests. The argument handling for `dart test` is all from `package:test` and the precompilation logic is built into `package:pub`, so we'd need to add some logic there to support this (assuming we can).

**Comment by jakemac53** [comment]
> I don't think there's a flag to disable the automatic `pub get` since `dart test` triggers some precompilation steps before running tests.

As in the precompilation of the test package itself? If pub still has a `run` command, you should be able to just do `pub run` instead of `pub get` (it will also trigger the precompilation).

**Comment by bkonyi** [comment]
> > I don't think there's a flag to disable the automatic `pub get` since `dart test` triggers some precompilation steps before running tests.
> 
> As in the precompilation of the test package itself? If pub still has a `run` command, you should be able to just do `pub run` instead of `pub get` (it will also trigger the precompilation).

Yes, precompilation of the test package. `dart test` doesn't call `pub run` or `pub get` directly, it simply calls [`getExecutableForCommand('test:test')`](https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/src/commands/test.dart#L64) and passes the returned executable path back for the VM to run.

**Comment by jakemac53** [comment]
> Yes, precompilation of the test package. `dart test` doesn't call `pub run` or `pub get` directly, it simply calls [`getExecutableForCommand('test:test')`](https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/src/commands/test.dart#L64) and passes the returned executable path back for the VM to run.

It looks like that calls [ensureUpToDate](https://github.com/dart-lang/pub/blob/9adca58e4fa8e1d94924e64a184532231826496c/lib/src/executable.dart#L323) - which only does a `pub get` if necessary, which is probably a _good_ thing in this case?

**Comment by bkonyi** [comment]
> > Yes, precompilation of the test package. `dart test` doesn't call `pub run` or `pub get` directly, it simply calls [`getExecutableForCommand('test:test')`](https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/src/commands/test.dart#L64) and passes the returned executable path back for the VM to run.
> 
> It looks like that calls [ensureUpToDate](https://github.com/dart-lang/pub/blob/9adca58e4fa8e1d94924e64a184532231826496c/lib/src/executable.dart#L323) - which only does a `pub get` if necessary, which is probably a _good_ thing in this case?

Do we want to be running `pub get` at all on test infrastructure or are we relying on a custom package config to reference `third_party/` dependencies and other local packages like we do in the SDK?

**Comment by jakemac53** [comment]
> Do we want to be running `pub get` at all on test infrastructure or are we relying on a custom package config to reference `third_party/` dependencies and other local packages like we do in the SDK?

The engine (at least for the most part) uses pub workspaces, with overrides for the third party deps (pulled in via a DEPS file IIRC). https://github.com/flutter/engine/blob/main/pubspec.yaml. This means `pub get` is safe to do.

**Review by jtmcdole — APPROVED** [review]
LGTM

**Review by jakemac53 — APPROVED** [review]

File: build/dart/internal/gen_executable_call.py
Comment (cbracken) [inline build/dart/internal/gen_executable_call.py:37]: Could save yourself a variable with `pushd "$CD_PATH" > /dev/null` and `popd > /dev/null`

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

File: build/dart/internal/gen_executable_call.py
Comment (matanlurey) [reply build/dart/internal/gen_executable_call.py]: Done, that's much simpler!

Added a PWD "test" as well.

**Comment by matanlurey** [comment]
I think @zanderso wanted a chance to weigh in here, and is OOO, so I'll check back with Monday before merging.

**Comment by zanderso** [comment]
Sorry, a couple more questions.

Is there any way to make the implied `pub get` run in `--offline` mode the same way that we do when setting up the tree in `tools/pub_get_offline.py`? I think this is not critical since the workspace *should* already be set up after a `gclient sync`, but I know folks forget to run `gclient sync` after doing a `git pull` all the time, and I'd like to avoid folks (and CI) getting their tree into a weird state.

Does `dart test` write to the source tree at all? If so, any files it writes will have to be cleaned up after the tests run, or we'll have to find somewhere else for it to write the files. For example, does it write any intermediate compilation artifacts anywhere? If it needs to do that, we'll have to find some way to avoid those being written in the source tree, or clean them up afterwords.  Stray files in the source tree risk confusing the caching of the source tree that CI does.

**Comment by matanlurey** [comment]
> Is there any way to make the implied `pub get` run in `--offline` mode the same way that we do when setting up the tree in `tools/pub_get_offline.py`? I think this is not critical since the workspace _should_ already be set up after a `gclient sync`, but I know folks forget to run `gclient sync` after doing a `git pull` all the time, and I'd like to avoid folks (and CI) getting their tree into a weird state.

No, there is not. We could definitely request one if we end up thinking its important.

But, on the importance, I'm struggling a bit to understand what we're trying to avoid:

- With workspace (and the enforcement of `pub_get_offline.py`) I'm not sure what it could do?
- We [already are using `dart test`](https://github.com/flutter/engine/blob/c4784aa7eadeead0005128922b32ffbc37717b66/testing/run_tests.py#L916-L918), both locally and on CI.

> Does `dart test` write to the source tree at all?

It does place files in `.dart_tool/test`: 

Screenshot 2024-09-28 at 6 47 10 PM

But I'm guessing since it's in `.gitignore` we aren't considering that the source tree?

**Comment by zanderso** [comment]
> * With workspace (and the enforcement of `pub_get_offline.py`) I'm not sure what it could do?

If someone does a `git pull` without doing a `gclient sync`, then `pub_get_offline.py` won't have run. So my question is more, what is the failure mode if someone does `git pull` and then immediately `dart test` in a situation where the `git pull` updated the `pubspec.yaml`. Will it be easy to debug? Would an `--offline` flag to `dart test` make the failure more obvious? To be clear these questions are not intended to block this PR, but they are things we should be thinking about.

> It does place files in `.dart_tool/test`:
> But I'm guessing since it's in `.gitignore` we aren't considering that the source tree?

CI does filesystem level caching, so even files that are `.gitignore`d in the source tree may persist between CI runs. Do we understand what the behavior of `dart test` is w.r.t. the files under `.dart_tool/test` when e.g. the version of `package:test` or the Dart SDK is updated?

Also a couple of questions for @bkonyi:
1. Does `dart test` send telemetry? Should we be explicitly disabling that on CI, somehow? Does `dart test` take a `--disable-telemetry` flag?
2. My understanding is that at some point in the near future, the Dart CLI will spawn a persistent compilation daemon. That means that these `dart test` invocation will spawn that process. Especially on CI, how will we be able to ensure that the daemons are killed between runs?

**Comment by jtmcdole** [comment]
> CI does filesystem level caching, so even files that are .gitignored in the source tree may persist between CI runs

That's a surprising tidbit - is it documented somewhere?  I would have imagined block level caching.

**Comment by bkonyi** [comment]
> Also a couple of questions for @bkonyi:
> 
> 1. Does `dart test` send telemetry? Should we be explicitly disabling that on CI, somehow? Does `dart test` take a `--disable-telemetry` flag?

The Dart CLI has a top-level `--suppress-analytics` flag that could be provided (e.g., `dart --suppress-analytics test`), but `package:unified_analytics` should be smart enough to detect if we're running on CI and disable telemetry automatically (detection logic found [here](https://github.com/dart-lang/sdk/blob/main/pkg/telemetry/lib/telemetry.dart#L14)).

> 2. My understanding is that at some point in the near future, the Dart CLI will spawn a persistent compilation daemon. That means that these `dart test` invocation will spawn that process. Especially on CI, how will we be able to ensure that the daemons are killed between runs?

@derekxu16 will know more about this since he's been the one working on it.

Would we want to keep the compilation server running and shared between tests on a shard? If not, there should be a way for the compilation server to have the same lifespan as the test process. If we do want to shared the compilation server, it should be as simple as adding a step to run `dart compilation-server shutdown` after all the tests have finished running.

**Comment by jakemac53** [comment]
> Do we understand what the behavior of `dart test` is w.r.t. the files under `.dart_tool/test` when e.g. the version of `package:test` or the Dart SDK is updated?

It is using the frontend server and passing a previous dill file to it to initialize from. The frontend server handles invalidating that as needed afaik (but if it doesn't do that, we have larger problems).

**Comment by zanderso** [comment]
> > CI does filesystem level caching, so even files that are .gitignored in the source tree may persist between CI runs
> 
> That's a surprising tidbit - is it documented somewhere? I would have imagined block level caching.

The recipes use https://flutter.googlesource.com/recipes/+/refs/heads/main/recipe_modules/cache/api.py, which reads/writes files to the CAS service.

**Comment by derekxu16** [comment]
> Is it possible to ensure that no other process connects to a compilation server?

Compilation servers will be associated with "info files", so it will be possible to specify which compilation server an invocation of `dart test` will connect to with the [`--resident-compiler-info-file` CLI option](https://github.com/dart-lang/sdk/blob/2eb6582c422616d1d8edef5a36be61cef7fc08a9/pkg/dartdev/lib/src/commands/compilation_server.dart#L22). So passing a unique `--resident-compiler-info-file` argument to each invocation of `dart test` will result in those invocations being hermetic.

**Comment by matanlurey** [comment]
Let's wrap up the discussion on this PR about possible semantics of tools.

I opened 2 more issues to encourage discussion elsewhere:
- Dirtying the source tree semantics [#155949](https://github.com/flutter/flutter/issues/155949)
- `dart test` semantics on engine CI [#155950](https://github.com/flutter/flutter/issues/155950)

**Comment by matanlurey** [comment]
@johnmccutchan or @zanderso Do either of you have feedback specifically about this PR (i.e. should block merging)? If not, I plan to merge this shortly.

**Comment by zanderso** [comment]
@matanlurey LGTM w/ `--suppress-analytics` passed to `dart test`. 

@derekxu16 @bkonyi Please keep the engine team and @johnmccutchan in the loop w.r.t. the persistent compilation daemon on CI.

**Pull request merged**
